data-prepper-core/src/main/java/org/opensearch/dataprepper/acknowledgements/AcknowledgementSetMonitor.java

```diff
@@ -0,0 +1,92 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.acknowledgements;
+
+import org.opensearch.dataprepper.model.event.EventHandle;
+import org.opensearch.dataprepper.event.DefaultEventHandle;
+import org.opensearch.dataprepper.model.acknowledgements.AcknowledgementSet;
+
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.Set;
+import java.util.HashSet;
+
+public class AcknowledgementSetMonitor implements Runnable {
+    private final Set<AcknowledgementSet> acknowledgementSets;
+    private final ReentrantLock lock;
+    private final int waitTime;
+    private AtomicBoolean doShutdown;
+
+    private DefaultAcknowledgementSet getAcknowledgementSet(final EventHandle eventHandle) {
+        return (DefaultAcknowledgementSet)((DefaultEventHandle)eventHandle).getAcknowledgementSet();
+    }
+
+    public AcknowledgementSetMonitor(int waitTime) {
+        this.acknowledgementSets = new HashSet<>();
+        this.lock = new ReentrantLock(true);
+        this.waitTime = waitTime;
+        this.doShutdown = new AtomicBoolean(false);
+    }
+
+    public void add(AcknowledgementSet acknowledgementSet) {
+        lock.lock();
+        acknowledgementSets.add(acknowledgementSet);
+        lock.unlock();
+    }
+
+    public void acquire(final EventHandle eventHandle) throws RuntimeException {
+        DefaultAcknowledgementSet acknowledgementSet = getAcknowledgementSet(eventHandle);
+        lock.lock();
+        boolean exists = acknowledgementSets.contains(acknowledgementSet);
+        lock.unlock();
+        // if acknowledgementSet doesn't exist then it means that the
+        // event still active even after the acknowledgement set is
+        // cleaned up.
+        if (exists) {
+            acknowledgementSet.acquire(eventHandle);
+        } else {
+            throw new RuntimeException("AcknowledgementSet does not exist");
+        }
+    }
+
+    public void release(final EventHandle eventHandle, final boolean success) throws RuntimeException {
+        DefaultAcknowledgementSet acknowledgementSet = getAcknowledgementSet(eventHandle);
+        lock.lock();
+        boolean exists = acknowledgementSets.contains(acknowledgementSet);
+        lock.unlock();
+        // if acknowledgementSet doesn't exist then it means some late
+        // arrival of event handle release after the acknowledgement set
+        // is cleaned up.
+        if (exists) {
+            boolean b = acknowledgementSet.release(eventHandle, success);
+        } else {
+            throw new RuntimeException("AcknowledgementSet does not exist");
+        }
+    }
+
+    // For testing
+    public int getSize() {
+        return acknowledgementSets.size();
+    }
+
+    public void shutdown() {
+        doShutdown.set(true);
+    }
+
+    @Override
+    public void run() {
+        try {
+            while (!doShutdown.get()) {
+                Thread.sleep(this.waitTime);
```

Comment:
I'm not sure what the waitTime will be. But, it might be more appropriate to run on a schedule rather than hold a thread.